Allow storage of multiple providers info

kliu57 opened the pull request:
Closes #383 

Save all newly inputted provider info to associative array settings.providers (stored in localStorage in providers)

Existing functionality is unchanged. We are only storing to settings.providers, not reading from settings.providers yet.

**Code changes:**

- `src/lib/ChatCraftProvider.ts` — new class which stores information for a provider such as provider name, url, key.
- `src/lib/settings.ts` — added the variable `providers` to Settings, which is an associative array of `ChatCraftProvider` objects
- `src/components/Message/AppMessage/Instructions.tsx` — added code so when new ChatCraft users enters a key and clicks the Save button, the provider info is stored to `settings.providers`
- `src/components/PreferencesModal.tsx` — added code to the event handlers of the API URL field, API Key field, Remove button, so that in addition to performing their usual behaviour, they also store/remove data from settings.providers

**Behaviour of the new code:**

1) New users to ChatCraft — the moment they input their api key, the provider information will be stored in settings.providers. You can test this by checking localStorage.

2) Existing users to ChatCraft — their current api key will not yet be stored in settings.providers. Will be addressed in release 1.2.

3) When a user changes their provider url in Settings — the moment they toggle to another API url, settings.providers is cleared, to match the current behaviour. This behaviour will change in release 1.2 (In 1.2 we will not clear the providers and instead keep them stored in settings.providers to be read from at a later date)

4) When a user enters a new/different provider key in Settings — the moment they enter a key, the key stored in settings.providers is stored or overwritten. There will never be more than one key stored for openai, for example.

5) When a user clicks the Remove button in Settings — if the current provider is stored in settings.providers it is removed from settings.providers.

![image1](https://github.com/tarasglek/chatcraft.org/assets/98062538/867b828a-fa2b-4caa-94ab-9d97d035b7e7)

**How to check localStorage in Chrome DevTools:**

Right click + Inspect (or F12)
![image2](https://github.com/tarasglek/chatcraft.org/assets/98062538/bc072cbf-967e-41cb-9f38-f53c4e0f17c9)

kliu57 commented:
@humphd You asked for the UI in the settings to iterate over the list of providers; however right now my settings.providers is an associative array that only stores the providers the user has given (so it could only contain 1). We would want to show them 2 providers in the dropdown and not 1. Thoughts?

![image](https://github.com/tarasglek/chatcraft.org/assets/98062538/17567f00-d369-43b4-9e9a-4d7988a10ee4)

humphd commented on src/components/PreferencesModal.tsx:
```diff
@@ -200,7 +200,20 @@ function PreferencesModal({ isOpen, onClose, finalFocusRef }: PreferencesModalPr
                 value={settings.apiKey || ""}
-                onChange={(e) => setSettings({ ...settings, apiKey: e.target.value })}
+                onChange={(e) => {
+                  const newProvider =
```
I wouldn't do this logic in the UI code, move it to the `ChatCraftProvider` class via a `static` method named `fromUrl()`, which parses a URL into a provider instance:

```ts
const newProvider = ChatCraftProvider.fromUrl(settings.apiUrl);
...
[newProvider.name]: newProvider,
```

humphd commented:
> @humphd You asked for the UI in the settings to iterate over the list of providers; however right now my settings.providers is an associative array that only stores the providers the user has given (so it could only contain 1). We would want to show them 2 providers in the dropdown and not 1. Thoughts?
> 
> ![image](https://private-user-images.githubusercontent.com/98062538/300152379-17567f00-d369-43b4-9e9a-4d7988a10ee4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYzNjQ0NzIsIm5iZiI6MTcwNjM2NDE3MiwicGF0aCI6Ii85ODA2MjUzOC8zMDAxNTIzNzktMTc1NjdmMDAtZDM2OS00M2I0LTllOWEtNGQ3OTg4YTEwZWU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI3VDE0MDI1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlYTBjZWEzMTRhZjA5NDJiZWJmNTBhMWJlMjYyOTc0YzVkM2E4NTM1NjI1NjkxYjE2ZTc4ODY4OWZkOGRiNGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.a6itpKYUgm9iTsKRaxIdamABjbg9EqjcPGx2GdOlIgM)

I think we have to separate the list of providers a user has configured in their settings from the list of providers we support.  In other words, we might support 3 different providers in the future, but a user might only have 1 or 2 configured.  So the UI can use the list of supported providers to show possible providers for a user to configure.

humphd commented:
I saw [Simon tweeting yesterday](https://twitter.com/simonw/status/1751276501809647919?t=Q-jSo3oqjJCmqVa-pXrlSQ) about this same thing (using OpenAI compatible providers), and people mention:

- [TogetherAI](https://www.together.ai/products#inference)
- [Anyscale](https://www.anyscale.com/endpoints)
- [Fireworks.ai](https://readme.fireworks.ai/docs/quickstart)

Might be interesting to support/try these with ChatCraft.

humphd commented on src/lib/ChatCraftProvider.ts:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  // Parses url into instance of ChatCraftProvider
+  static fromUrl(url: string) {
```
What if you do `static fromUrl(url: string, apiKey: string) {` so that you require the `apiKey` at the same time, thus getting rid of the case where it's set to `""`.

humphd commented on src/lib/ChatCraftProvider.ts:
```diff
@@ -0,0 +1,44 @@
+    this.apiKey = key;
+  }
+
+  setUrl(url: string) {
```
I don't think `setUrl` or `setKey` are necessary, since you are exposing `apiUrl` and `apiKey` as public members.

humphd commented on src/lib/ChatCraftProvider.ts:
```diff
@@ -0,0 +1,44 @@
+
+export class ChatCraftProvider {
+  id: string;
+  name: string;
```
Let's narrow this type from `string` to the values we want:

```ts
type ProviderName = 'openai' | 'openrouter.ai';
...
name: ProviderName;
```

Amnish04 replied:
I was about to comment something similar.

Shouldn't we rather use an **enum** here so we don't have to hard code string values like below
![image](https://github.com/tarasglek/chatcraft.org/assets/78865303/d7e7c654-ce34-4466-bf99-0cdb85261e8d)

With an enum:
```
enum ProviderName {
  OPEN_AI = "OpenAI",
  OPEN_ROUTER = "OpenRouter.ai",
}

...
...

// Parses url into instance of ChatCraftProvider
static fromUrl(url: string) {
  let name = null;

  switch (url) {
    case OPENAI_API_URL:
      name = ProviderName.OPEN_AI;
      break;
    case OPENROUTER_API_URL:
      name = ProviderName.OPEN_ROUTER;
      break;
    default:
      throw new Error("Invalid URL");
  }

  return new ChatCraftProvider(name, url, "");
}
```

At this point, I feel like we could even have an `apiUrl` to `ProviderName` mapping. That could simplify this to\
```ts
// Parses url into instance of ChatCraftProvider
static fromUrl(url: string) {
  let name = null;

  if (mapping[url]) {
    const providerName = mapping[url];
    return new ChatCraftProvider(providerName, url, "");
  }
  else {
    throw new Error("Invalid URL");
  }
}
```

The decision depends on how we are our use case and vision for the future.

humphd commented on src/lib/ChatCraftProvider.ts:
```diff
@@ -0,0 +1,44 @@
+      default:
+        throw new Error("Invalid URL");
+    }
+
```
You should also do `fromJSON()` so we can parse this out of JSON and back into an instance.

humphd reviewed (changes requested):
This is getting better and better, nice work.  I've left a few thoughts for things I'd do differently.

Amnish04 commented:
@kliu57 I've been testing this for a while, and the first thing I was about to comment was that we shouldn't delete the existing `apiKey` when switching providers. But then I saw this
![image](https://github.com/tarasglek/chatcraft.org/assets/78865303/6be21473-cb82-4c33-bdfe-3cc85300c300)
So I am assuming that's not an issue.

Rest works well and as you've described!

Just left a [comment](https://github.com/tarasglek/chatcraft.org/pull/371#discussion_r1473678243) for handling `ProviderNames`. But I am not sure what would be the best approach in our case. Maybe @humphd can confirm.

kliu57 commented:
@humphd Made the following changes to `ChatCraftProvider`, addressing all your review comments
- changed `fromUrl` function to take `apiKey` param as well
- removed `setUrl` and `setKey` functions
- changed `name` member var of ChatCraftProvider from `string` data type to `ProviderName` where `type ProviderName = "OpenAI" | "OpenRouter.ai"`
- added `fromJSON` function

@Amnish04 I created an urlToNameMap object which can map the url to the name
```
const urlToNameMap: { [key: string]: ProviderName } = {
  [OPENAI_API_URL]: "OpenAI",
  [OPENROUTER_API_URL]: "OpenRouter.ai",
};
```

humphd commented on src/components/PreferencesModal.tsx:
```diff
@@ -164,9 +165,14 @@ function PreferencesModal({ isOpen, onClose, finalFocusRef }: PreferencesModalPr
+                    ...settings,
+                    apiUrl: e.target.value,
+                    apiKey: undefined,
+                    providers: {},
```
Question: are you losing data here?  Why clear all values vs. the one you're working on?

humphd reviewed (approved):
Looks good, just one question.

kliu57 commented:
> Looks good, just one question.

@humphd 

**Why I wanted to clear the key from settings.providers when the user toggled the API URL field:**

If I kept the key stored in settings.providers, the user may not know that we have it stored. For example they toggle from OpenAI to OpenRouter, and we kepts their OpenAI key stored in settings.providers. Say they toggle back to OpenAI, our UI shows a blank where the key is, even though we have their key stored in settings.providers. They would have no idea we stored their key and wouldn't have the ability to clear it from storage.

**Why I didn't I just clear the current key from settings.providers when the user toggled the API URL field (and instead set it to {} thus clearing all values ):**

I did this because in the next release, I won't even be clearing the key at all. The code to clear the key will be removed, so I figured it didn't matter if I cleared the current key by clearing the entire array, or I cleared the current key by removing it specifically. This code of clearing the key will be removed in the next release so that we can keep the stored provider when we toggle.

humphd commented:
Sounds good, I wanted to confirm we hadn't missed a case where we'd lose data without meaning to.  Thanks for clarifying.

Amnish04 commented:
> @humphd Made the following changes to `ChatCraftProvider`, addressing all your review comments
> 
>     * changed `fromUrl` function to take `apiKey` param as well
> 
>     * removed `setUrl` and `setKey` functions
> 
>     * changed `name` member var of ChatCraftProvider from `string` data type to `ProviderName` where `type ProviderName = "OpenAI" | "OpenRouter.ai"`
> 
>     * added `fromJSON` function
> 
> 
> @Amnish04 I created an urlToNameMap object which can map the url to the name
> 
> ```
> const urlToNameMap: { [key: string]: ProviderName } = {
>   [OPENAI_API_URL]: "OpenAI",
>   [OPENROUTER_API_URL]: "OpenRouter.ai",
> };
> ```
> 
> Tested `fromJSON()` function to make sure it can parse JSON to `ChatCraftProvider` object:
> 
> ![image](https://private-user-images.githubusercontent.com/98062538/301370791-8b8d1314-6280-4d44-ab21-d917c697cbcb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY4MTM2OTEsIm5iZiI6MTcwNjgxMzM5MSwicGF0aCI6Ii85ODA2MjUzOC8zMDEzNzA3OTEtOGI4ZDEzMTQtNjI4MC00ZDQ0LWFiMjEtZDkxN2M2OTdjYmNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAxVDE4NDk1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1ODI1MDJjNDQ2MjVmYTE1ZGM4NjQ1YjBmMGFhMTg2OTEzN2EzZjY4NjhhOGQ4YTM0YzBhMjVhYTY0ZjM0MDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WKpZZAYei2zeqAc9aBzV74giMLHnXamlsa7A1fzzVtI)

@kliu57 Thanks for the changes. I would name the mapping something like `providerNamesByUrl` to be more specific. But that's just my personal preference, and it's up to you if you want to change that.

Overall, looks good to me!
